marlowe-playground-client/src/MetadataTab/View.purs

```diff
@@ -126,11 +157,20 @@ metadataView metadataHints metadata metadataAction =
                   ]
               ]
           ]
-        , generateMetadataList _roleDescriptions _roles SetRoleDescription DeleteRoleDescription "Role" "role"
-        , generateMetadataList _choiceDescriptions _choiceNames SetChoiceDescription DeleteChoiceDescription "Choice" "choice"
-        , generateMetadataList _slotParameterDescriptions _slotParameters SetSlotParameterDescription DeleteSlotParameterDescription "Slot parameter" "slot parameter"
-        , generateMetadataList _valueParameterDescriptions _valueParameters SetValueParameterDescription DeleteValueParameterDescription "Value parameter" "value parameter"
+        , generateMetadataList _roleDescriptions _roles (onlyDescriptionRenderer SetRoleDescription DeleteRoleDescription) "Role" "role" (\x -> SetRoleDescription x mempty)
+        , generateMetadataList _choiceInfo _choiceNames choiceMetadataRenderer "Choice" "choice" (\x -> SetChoiceDescription x mempty)
+        , generateMetadataList _slotParameterDescriptions _slotParameters (onlyDescriptionRenderer SetSlotParameterDescription DeleteSlotParameterDescription) "Slot parameter" "slot parameter" (\x -> SetSlotParameterDescription x mempty)
+        , generateMetadataList _valueParameterDescriptions _valueParameters (onlyDescriptionRenderer SetValueParameterDescription DeleteValueParameterDescription) "Value parameter" "value parameter" (\x -> SetValueParameterDescription x mempty)
         ]
     )
   where
+  generateMetadataList ::
+    forall c.
+    Lens' MetaData (Map String c) ->
+    Lens' MetadataHintInfo (Set String) ->
+    (String -> c -> Boolean -> (MetadataAction -> a) -> String -> String -> Array (HTML p a)) ->
+    String ->
+    String ->
+    (String -> MetadataAction) ->
+    Array (HTML p a)
```

Comment:
I'm not sure how helpful/necessary this type signature is...? :confused:

Reply:
It is quite curious: it is actually required for compilation to work, because it needs to instantiate c in two different ways, so if you don't have the `forall c` as a declaration it will just give type mismatch. The `p` and the `a` refer to the type of the of the parent function, which is something I didn't know one could do until this